extensions-android/src/main/java/com/kevinmost/koolbelt/extension/android/ViewUtil.kt

```diff
@@ -37,52 +38,58 @@ fun <T : View> Dialog.find(@IdRes id: Int): T? {
 }
 
 fun ViewGroup.getChildren(): List<View> {
-  val list = ArrayList<View>(childCount)
-  for (index in 0..childCount) {
-    list.add(getChildAt(index))
-  }
-  return list
-}
-
-inline fun ViewGroup.forEachChild(forEachChild: View.() -> Unit) {
-  for (index in 0..childCount) {
-    getChildAt(index)?.let {
-      forEachChild(it)
+  return mutableListOfSize<View>(childCount).apply {
+    for (index in 0..childCount) {
+      getChildAt(index)?.let { add(it) }
     }
   }
 }
 
 /**
- * @return this view if it of type [T]; otherwise, the first child of this view that is of type [T].
- * The search is performed depth-first.
+ * NOTE: This [Sequence] could easily get unstable if you change the View hierarchy while it's in scope.
  */
-fun <T : View> View.getChildOfType(type: Class<T>): T? {
-  if (type.isInstance(this)) {
-    return type.cast(this)
-  }
-  if (this is ViewGroup) {
-    for (index in 0..childCount) {
-      getChildAt(index).getChildOfType(type)?.let { return it }
+fun ViewGroup.toSequence(): Sequence<View?> {
+  var currentIndex = 0
+  return generateSequence {
+    return@generateSequence try {
+      getChildAt(currentIndex)
+    } finally {
+      currentIndex++
     }
   }
-  return null
 }
 
-/**
- * @return a list of all children of this View that are of type [T]. This View is also included in
- * the list as the first element if it is of type [T]. The search is performed depth-first.
- */
-fun <T : View?> View.getAllChildrenOfType(type: Class<T>): List<T> {
-  val list = mutableListOf<T>()
-  if (type.isInstance(this)) {
-    list.add(type.cast(this))
+fun <T : View> View.firstDescendantOfInstance(type: Class<T>, includeSelf: Boolean = true): T? {
+  return when {
+    includeSelf && type.isInstance(this) -> type.cast(this)
+    this is ViewGroup -> toSequence()
+        .filterNotNull()
+        .map { child -> child.firstDescendantOfInstance(type, includeSelf = true) }
```

Comment:
Hmmm... I think this could be problematic for more complex views since even if it finds a match, it will continue visiting the other children.

Maybe we could try replacing a `.map { }` call with `.firstOrNull { }`?

```
  fun <T : View> View.firstDescendantOfInstance(type: Class<T>, includeSelf: Boolean = true): T? {
    return when {
      includeSelf && type.isInstance(this) -> type.cast(this)
      this is ViewGroup -> toSequence()
          .filterNotNull()
          .firstOrNull {
              child -> child.firstDescendantOfInstance(type, includeSelf = true) != null
          }?.let { type.cast(this) }
      else -> null
    }
  }
```

The only issue I have here is that the recursive call already casts when it returns, but the nature of `firstOrNull { }` means that we end up returning the sequence type (which is `View?`). We'd have to cast again, which isn't a huge deal but it's kinda ugly... there is probably a more elegant solution!

Let me know what you think!